pkg/cmd/auth/login/login.go

```diff
@@ -0,0 +1,284 @@
+				}
+
+				opts.Token = strings.TrimSpace(string(token))
+				if opts.Hostname == "" {
+					opts.Hostname = defaultHostname
+				}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return loginRun(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.Hostname, "hostname", "", "The hostname of the GitHub instance to authenticate with")
+	cmd.Flags().Bool("with-token", false, "If specified, token is read from STDIN")
+
+	return cmd
+}
+
+func loginRun(opts *LoginOptions) error {
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	if opts.Token != "" {
+		if opts.Hostname == "" {
+			panic("empty hostname would leak oauth_token")
+		}
+		err := cfg.Set(opts.Hostname, "oauth_token", opts.Token)
+		if err != nil {
+			return err
+		}
+		err = cfg.Write()
+		if err != nil {
+			return err
+		}
+
+		httpClient, err := opts.HttpClient()
+		if err != nil {
+			return err
+		}
+
+		apiClient := api.NewClientFromHTTP(httpClient)
+
+		hasScopes, _, err := apiClient.HasScopes(expectedScopes...)
+		if err != nil {
+			return fmt.Errorf("could not verify token: %w", err)
+		}
+
+		if !hasScopes {
+			return fmt.Errorf("token missing at least one of the required scopes: %v", expectedScopes)
+		}
+
+		return nil
+	}
+
+	isTTY := opts.IO.IsStdoutTTY() && opts.IO.IsStdinTTY()
+
+	if !isTTY {
+		return errors.New("token must be passed via STDIN and --with-token when unattached to TTY")
+	}
+
+	// TODO consider explicitly telling survey what io to use since it's implicit right now
+
+	hostname := opts.Hostname
+
+	if hostname == "" {
+		var hostType int
+		err := prompt.SurveyAskOne(&survey.Select{
+			Message: "What account do you want to log into?",
+			Options: []string{
+				"GitHub.com",
+				"GitHub Enterprise",
+			},
+		}, &hostType)
+
+		if err != nil {
+			return fmt.Errorf("could not prompt: %w", err)
+		}
+
+		isEnterprise := hostType == 1
+
+		hostname = defaultHostname
+		if isEnterprise {
+			err := prompt.SurveyAskOne(&survey.Input{
+				Message: "GHE hostname:",
+			}, &hostname, survey.WithValidator(survey.Required))
+			if err != nil {
+				return fmt.Errorf("could not prompt: %w", err)
+			}
+		}
+	}
+
+	fmt.Fprintf(opts.IO.ErrOut, "- Logging into %s\n", hostname)
+
+	var authMode int
+	err = prompt.SurveyAskOne(&survey.Select{
+		Message: "How would you like to authenticate?",
+		Options: []string{
+			"Login with a web browser",
+			"Paste an authentication token",
+		},
+	}, &authMode)
+	if err != nil {
+		return fmt.Errorf("could not prompt: %w", err)
+	}
+
+	if authMode == 0 {
+		_, err := config.AuthFlowWithConfig(cfg, hostname, "")
+		if err != nil {
+			return fmt.Errorf("failed to authenticate via web browser: %w", err)
+		}
+	} else {
+		fmt.Fprintln(opts.IO.ErrOut)
+		fmt.Fprintln(opts.IO.ErrOut, heredoc.Doc(`
+				Tip: you can generate a Personal Access Token here https://github.com/settings/tokens
+				The minimum required scopes are 'repo' and 'read:org'.`))
```

Comment:
I'm torn; on the one hand, printing to STDOUT is absolutely fine here (and survey is using STDOUT). On the other, we print this kind of stuff to STDERR in other commands and I'd like to be consistent.

Reply:
> On the other, we print this kind of stuff to STDERR in other commands and I'd like to be consistent.

True. But the way I see it, the only reason we separate out "special" notices from regular output is because we expect that the user might be piping regular output to a file or a script. If the user is not doing that (for this command, they are not likely to), there is no incentive for us to print notices on a separate stream.

Reply:
Some software writes to the tty in raw mode for these types of informational notices, so the text is neither in `stdout` nor `stderr`.